fix #8859, add NitroKey3 HOTP support

UndeadDevel opened the pull request:
This PR:
- fixes https://github.com/QubesOS/qubes-issues/issues/8859
- adds support for NitroKey3 (HOTP mode)
- updates comments to reflect NK3 support
- changes some `yk-*` file names to remove `yk-` as they are now for both YK and NK3
- adds new dependencies (`coreutils` for `base32`, `util-linux` for `hexdump`, `xxd` or `vim-common` for `xxd`, all available by default in dom0) to rpm file
- bumps version to 1.2.0

I have tested the new NK3 functionality, but don't have a YubiKey to test with; since the YK functionality is not modified aside from @marmarek 's tip in the above linked issue, it should still work / work again, since that issue broke YK support.

Should this PR be accepted, I will provide another one to update the docs.

UndeadDevel committed:
add NitroKey3 HOTP support

UndeadDevel committed:
fix dependencies

UndeadDevel commented:
Hang on, it looks like there's a bug with the HOTP code for a small number of outputs...fixing...

UndeadDevel committed:
fix hotp issue for some outputs

UndeadDevel commented:
Fixed; ready for review.

UndeadDevel committed:
simplify shell reference

marmarek commented:
> changes some `yk-*` file names to remove `yk-` as they are now for both YK and NK3

This part isn't that easy - it still need to handle old names for compatibility. Otherwise, the update will break login for some users, which is not acceptable.

marmarek commented on bin/yk-auth:
```diff
@@ -28,13 +28,50 @@ if [ -n "$pass_hash" ]; then
+	hmac=$(printf "$(printf "%016X" "$counter" | sed -e 's/\(..\)/\\x\1/g')" | 
+		openssl dgst -sha256 -mac HMAC -macopt "hexkey:$key" | sed -e 's/^.* //')
+	otp=$(printf "%08d" "$(( ( (0x${hmac:$((2 * 0x${hmac: -1})):8}) & 0x7FFFFFFF) % (10**8) ))" )
+	response=$(qvm-run -a --nogui -p $vm "\$HOME/.local/bin/nitropy nk3 secrets get loginxs" | tail -n1 | tr -d "\n")
```
How is `nitropy` installed? I take there is no proper distribution package, right? Unfortunately, pip (or pipx for that matter) has rather poor integrity validation and is something we do not recommend using. I see there is also an option to download binary (binary? isn't that python?) directly from https://github.com/Nitrokey/pynitrokey/releases/tag/v0.4.44, but it isn't signed either.
Lets see, maybe this can be fixed: https://github.com/Nitrokey/pynitrokey/issues/495

UndeadDevel replied:
Yes, `pipx` seems to be the main way for the typical end user.

The "binary" is actually an executable Linux binary; unpacking it and doing `sudo ln -s /path/to/unpacked/binary /usr/local/bin/nitropy` allows easy execution without calling python3; it may still use python3 "under the hood", but who knows; when I uninstall python3 in a dvm (as far as possible; the process crashes because qubes packages need python3) and via `qvm-console-dispvm` unpack and execute the binary, it works.
Edit: any crypto ops currently don't work with the binary, however, due to [this issue](https://github.com/Nitrokey/pynitrokey/issues/431)

Signed artifacts or even better official distro packaging would be nice, yes.

marmarek commented on bin/yk-auth:
```diff
@@ -1,24 +1,24 @@
     # password to logs, even with set -x
-    pass_hash_clear=$(cat $CONFIG_DIR/yk-login-pass | grep -v '^#' | tr -d '\n' |
-        openssl dgst -sha1 -r | cut -f1 -d ' ')
+    pass_hash_clear=$(cat $CONFIG_DIR/login-pass | grep -v '^#' | tr -d '\n' | openssl dgst -sha1 -r | cut -f1 -d ' ')
```
Please don't unfold long lines, they should be (approximately) 80 chars max.

marmarek commented on bin/yk-auth:
```diff
@@ -28,13 +28,50 @@ if [ -n "$pass_hash" ]; then
+
+# to use qubes-yubikey-dom0 in a configuration compatible with a NitroKey3, execute the following in dom0:
+# echo -n "your-20-digit-secret" | base32 | sudo tee /etc/qubes/yk-keys/nk-hotp-secret
+[ -r "$CONFIG_DIR/nk-hotp-secret" ] && secret=$(cat $CONFIG_DIR/nk-hotp-secret | grep -v '^#' | tr -d '\n')
```
```suggestion
[ -r "$CONFIG_DIR/nk-hotp-secret" ] && nk_secret=$(cat $CONFIG_DIR/nk-hotp-secret | grep -v '^#' | tr -d '\n')
```
and update variable usage below too

UndeadDevel committed:
implement suggestions

UndeadDevel commented:
@marmarek 
> This part isn't that easy - it still need to handle old names for compatibility. Otherwise, the update will break login for some users, which is not acceptable.

Added checks for old filenames in the script. Edit: ~~or does the update create the new files additionally? In that case I'd need to modify the checks.~~ improved the check for old filenames such that even if the update adds those files, the old ones will still work.

All other suggestions also implemented and I've expanded the command to also work with a proper `nitropy` installation (currently not possible due to lacking distro package and the available binary not supporting crypto-ops due to [this issue](https://github.com/Nitrokey/pynitrokey/issues/431)).

UndeadDevel committed:
improve check for old filenames

UndeadDevel committed:
fix broken login after boot

marmarek commented:
Ok, now it looks mostly fine. Please run shellcheck on the script, there are few pre-existing issues, but your change add several more (especially missing quotes in quite a few places).

marmarek commented:
PipelineRetry

UndeadDevel committed:
shellcheck

UndeadDevel commented:
Implemented / fixed all `shellcheck` suggestions / warnings except for the ones on line 86 and 88.
For line 86: 
```
hmac=$(printf "$(printf "%016X" "$counter" | sed -e 's/\(..\)/\\x\1/g')" | 
              ^-- SC2059 (info): Don't use variables in the printf format string. Use printf '..%s..' "$foo".
```
If I put `'..%s..'` or `'%s'` or the same with double quotes there, it will produce a different hmac and thus a different (wrong) OTP...no idea why. In the [original script](https://github.com/bfreis/otp/blob/master/otp) there is also no format argument used.

For line 88:
```
otp=$(printf "%08d" "$(( ( (0x${hmac:$((2 * 0x${hmac: -1})):8}) & 0x7FFFFFFF) % (10**8) ))" )
                              ^-- SC3057 (warning): In POSIX sh, string indexing is undefined.
                                              ^---------^ SC3057 (warning): In POSIX sh, string indexing is undefined.
                                                                                   ^-- SC3019 (warning): In POSIX sh, exponentials are undefined.
```
These are "just" warnings, but it's beyond my skill to fix...the code is the basically the same as in the above mentioned script; I've refactored it a bit, but my changes are not the cause of the warnings.

marmarek commented:
> hmac=$(printf "$(printf "%016X" "$counter" | sed -e 's/\(..\)/\\x\1/g')" |

What about using xxd instead of printf to decode hex? Something like:
```
hmac=$(printf "%016X" "$counter" | xxd -r -ps | openssl ...
```
(sed is not necessary then)

> ^-- SC3057 (warning): In POSIX sh, string indexing is undefined.

This looks to be bash-specific feature, maybe simply change shebang at the top to `#!/bin/bash`

marmarek commented:
BTW should this work with Nitrokey Pro? I think it does support HOTP too.

UndeadDevel committed:
fix remaining shellcheck problems

UndeadDevel commented:
It worked, `shellcheck` is clean now!
Regarding the failed pipeline, it seems to be due to
```
09:09:56,781 [executor:docker:5a19419cfb15] output:   Running scriptlet: qubes-release-4.2-2.fc37.noarch                      60/60
09:09:56,781 [executor:docker:5a19419cfb15] output: System has not been booted with systemd as init system (PID 1). Can't operate.
09:09:56,782 [executor:docker:5a19419cfb15] output: Failed to connect to bus: Host is down
09:09:56,782 [executor:docker:5a19419cfb15] output: warning: %posttrans(qubes-release-4.2-2.fc37.noarch) scriptlet failed, exit status 1
09:09:56,782 [executor:docker:5a19419cfb15] output: 
09:09:56,783 [executor:docker:5a19419cfb15] output: Error in POSTTRANS scriptlet in rpm package qubes-release
```
...so not caused by this PR I think.

>BTW should this work with Nitrokey Pro? I think it does support HOTP too.

Theoretically NK3, NKPro and even NK Storage should be compatible, but I don't have the latter two to test with and I'm not sure how one would tell them via command line (`qvm-run`) to release their OTP...in `nitropy pro` or `nitropy storage` there is no option to set it up, only in `nitropy nk3`. However, with my NK3 I can do FIDO2 stuff in `nitropy fido2`, even though that's supposed to be for NitroKey FIDO2 devices, so maybe the same `nitropy nk3` commands would also work with NKPro and NK Storage, but again, I can't test it; there is a forum topic on the NK forums about this, however, maybe it will generate an answer.

UndeadDevel committed:
add xxd to dependencies

UndeadDevel commented:
Hmm, looks like `xxd` is not available on Fedora 32 for Qubes 4.1 (Edit: or it's part of another package, e.g. `vim-common`?), but then again I haven't tested `qubes-dom0-yubikey` on 4.1, maybe it worked there; in any case, if this PR is eventually accepted I guess it should only be published as an update for 4.2...as for the pipeline for 4.2 I don't know why it keeps failing, but it could be dependency issues regarding the `libvirt-daemon` packages...doesn't seem related to this PR but a problem with the testing environment.

marmarek commented:
> Hmm, looks like `xxd` is not available on Fedora 32 for Qubes 4.1 (Edit: or it's part of another package, e.g. `vim-common`?),

You can put something like:
```
%if 0%{?fedora} < 36
Requires: vim-common
%else
Requires: xxd
%endif
```
(I checked - it's f36 where the package was split)

as for the other failure, unrelated CI issue, should be fixed in the meantime

UndeadDevel committed:
fix dependencies

UndeadDevel commented:
Well, still no answer about NitroKey Pro or Storage on their forum and the latest `nitropy` release (released after you opened the issue about detached signatures) still doesn't have a detached signature. It seems this is as good as it gets for now. In any case, at least for the detached signature issue, the change would be in the documentation, not this PR.

All other problems are resolved.

UndeadDevel commented:
Okay so someone who seems pretty knowledgeable on the NitroKey forum (a member) says `nitropy` doesn't support HOTP on the NK Pro and NK Storage; the official support is only via the GUI-only NitroKey App. There is, however, a [third party CLI tool](https://github.com/d-e-s-o/nitrocli) that could be used, but it's not from NitroKey and, again, I don't have the Pro or Storage to test with, so I'm reluctant to try to "wing it" by somehow including support for it and those two devices.

marmarek commented:
Okay, thanks for checking

marmarek commented:
I can confirm the YubiKey support still works :)

UndeadDevel commented:
Thanks for merging! ~~I'll whip up another PR for the docs in the coming days.~~

Edit: I'll wait ~2 weeks with the doc PR as it's too complicated right now to setup `nitropy` with gpg verification what with the mentioned [issue with oscrypto](https://github.com/Nitrokey/pynitrokey/issues/431), [lacking release asset signatures](https://github.com/Nitrokey/pynitrokey/issues/495) and [lacking signature on the pypi release](https://files.pythonhosted.org/packages/fe/5a/2ce54bd60ba43dee7ed8f909c3f997a9e03517416523213e4f7bed480f3a/pynitrokey-0.4.45.tar.gz.asc).
Hopefully the situation will improve in the meantime.

UndeadDevel commented:
Update: none of the above mentioned issues have been resolved and I can't get `pip` or `pipx` to install the local, gpg-verified source package because for some reason it won't accept the presence of the gpg-verified system package `flit_core` dependency (Fedora 39 `flit` package).
I'll wait two more weeks, but if there is still no resolution I'll make the doc PR with just a regular `pipx` install without gpg verification, so that it's at least documented (though the code of this here PR also self-documents the main points); I can put a big warning about lacking verification that can hopefully be removed in the future once `nitropy` is either properly packaged or at least a signed release asset exists.

RoosDev commented:
Hello,

I m new user of QubesOs.
I set my yubikey like explain in the documentation.
I add the line : auth include yubikey
In the 2 next files :
/etc/pam.d/xscreensaver
/etc/pam.d/lightdm

But since I wrote it in lightdm , when I write my password on the lockscreen it tells me that my password is incorrect and I can t use my yubikey.

Does someone have an idea to fix this point ?

But i m afraid that I should format my disk...

Thanks

marmarek commented:
Don't post questions in a random pull request, see https://www.qubes-os.org/support/

RoosDev commented:
It s not post in random pull request,  I think there is a link between both, and thé problem is the same. But it could be my mistake.
